### ISSUE TITLE
Avoid applicative style

### ISSUE THREAD

**Issue by tfausak** [EdutainmentLIVE/haskell-style-guide#29]
``` hs
-- bad
User <$> getName <*> getEmail

-- good
name <- getName
email <- getEmail
pure (User name email)

-- (better)
pure User { userName = name, userEmail = email }
```

I thought I already had an issue about this, but I guess not. The applicative style is definitely compact, but it's hard to refactor or debug. Also it requires using constructors with positional arguments rather than names.

**Comment by tfausak** [comment]
Note that in general following this recommendation would mean giving up these operators:

``` hs
(<$>) :: Functor     f =>   (a -> b) -> f a -> f b
(<$)  :: Functor     f =>    a       -> f b -> f a
($>)  :: Functor     f => f  a       ->   b -> f b
(<*>) :: Applicative f => f (a -> b) -> f a -> f b
(<*)  :: Applicative f => f  a       -> f b -> f a
(*>)  :: Applicative f => f  a       -> f b -> f b
```

I'm okay with that, but I'd like to hear what the rest of the team thinks.

**Comment by gera-cameron** [comment]
Looks good to me! I'm cool with the do notation over the applicative. I think it gives us a more concrete concept of what's actually happening. 👍

**Comment by tfausak** [comment]
One thing that might make the monadic style more palatable is enabling [the `NamedFieldPuns` extension](https://downloads.haskell.org/~ghc/8.4.2/docs/html/users_guide/glasgow_exts.html#record-puns). (See #13 for a more general discussion of language extensions.)

``` hs
userName <- getName
userEmail <- getEmail
pure User { userName, userEmail }
```

For what it's worth, those same puns can also be used for binders.

``` hs
greet :: User -> String
greet User{userName} = "Hello, " ++ userName ++ "!"

**Comment by tfausak** [comment]
My main beef with the applicative style has to do with constructing records, which I think is how we usually use it. Given the following record and helpers:

``` hs
data User = User { userName :: String, userEmail :: String }
getName :: Applicative f => f String
getEmail :: Applicative f => f String
```

You can pretty easily get in a situation where the types line up but they don't mean the right thing. 

``` hs
User <$> getEmail <*> getName
```

The obvious solution is to give names and emails distinct types, but I don't think it's practical to require every field of every record to be unique. 

If you could use the applicative style with named record fields, I think I'd be okay with it. Sadly there's no way to do that, short of providing your own useless lambda that constructs the record.

``` hs
(\ name email -> User { userName = name, userEmail = email) <$> getName <*> getEmail
```

One of the things that's an actual upside of the monadic style (as opposed to a downside of the applicative style) is the ease of refactoring and debugging. Given this contrived `User` example we've got, it's really easy to change the monadic style to do some logging and additional checks.

``` hs
userName <- getName
traceM $ "Got user name: " ++ show userName
when (null userName) $ fail "User name cannot be empty!"
```

Doing the same thing with an applicative expression is painful.

**Comment by tfausak** [comment]
@qxjit and I just chatted about this. I've changed my mind: We'll use applicative style when it's appropriate. Some reasons:

- Some things can only provide `Applicative` interfaces. For example, Orville's table mapper can't provide a `Monad` instance, so we can't write it with `do` notation. 
- If we want to provide (for example) better JSON errors, monadic instances limit us to the first problem. Applicative instances could give arbitrarily many errors at once. 
- We pretty much use `newtype` wrappers for each record field anyway. So the danger of messing up the order of a record field is pretty low. 

Some fallout:

- Do we want to use `<$>` instead of `fmap` in general? I still prefer `fmap` over `<$>` (and functions over operators in general). But if an expression could reasonably be extended with `<*>`, I think it makes sense to use `<$>`. 
- Just because something can be written in the applicative style doesn't mean that it should be. This is similar to how using `case` can be clearer than using `maybe`. Use your discretion; don't go crazy with the operators.
- Do we want to use `<$`, `$>`, `<*`, and `*>`? In my experience they don't come up too often outside of parsers.

**Comment by telser** [comment]
I love the discussion here!

My thoughts on the questions posed:

- Will say that I _strongly_ prefer `fmap` over `<$>` in general, but concede that if using `<*>` for an expression, then using `<$>` isn't the end of the world.

- I would argue against the other operators, _unless_ a parser is what you're working on and maybe not even then. Namely I'd be fine with disallowing them via hlint.

Now, supposing no other issues, would `ApplicativeDo` change how anyone feels about this? I haven't looked to see what the situation around that extension is currently, but if it isn't breaking anything else, would it not give us the ability to provide better errors and the readability advantages discussed here?

**Comment by tfausak** [comment]
> if using <*> for an expression, then using <$> isn't the end of the world

👍 

> I'd be fine with disallowing [the other operators] via hlint

👍 

I love the idea of `ApplicativeDo`, but for some reason I perceive it as kind of rickety. That's not based on anything though. It's been around for a few major GHC releases by this point, so it's probably safe to rely on 😆 

From my point of view, the combination of `ApplicativeDo` and `NamedFieldPuns` is pretty great. We get to write monadic-looking code with `do` notation that desugars into a simple applicative expression, and we don't have to use any operators. 

The only potentially weird thing is that attempting to modify a `do` block might fail. For example:

``` hs
example :: Applicative something => something User
example = do
  userName <- getName
  traceShowM userName -- this is fine
  -- when (null userName) (fail "oh no!") -- but this needs `Monad something => ...`
  pure User { userName }
```

In other words, all `do` blocks are equal, but some are more equal than others.

**Comment by tfausak** [comment]
Resurrecting this discussion, it is possible to use the applicative style without using applicative operators. I don't know if it's actually better, but I think it's worth considering. Here's the code: https://gist.github.com/tfausak/3e5c178bfc50fe78e7c020b086fe7c92

And here's an example:

``` hs
data User = User { name, email :: String }
instance FromJSON User where

  -- before
  parseJSON = withObject "User" $ \ object -> User
    <$> object .: "name"
    <*> object .: "email"

  -- after
  parseJSON = withObject "User" $ \ object -> buoy User
    (object .: "name")
    (object .: "email")
```

**Comment by codygman** [comment]
> * Some things can only provide `Applicative` interfaces. For example, Orville's table mapper can't provide a `Monad` instance, so we can't write it with `do` notation.

@tfausak what about `ApplicativeDo`?

**Comment by tfausak** [comment]
Closing this one. We use applicative style all over the place and it's fine.

**Comment by aschmois** [comment]
Since we have `NamedFieldPuns` should we try and enforce it a bit more with that? I still feel like applicative style is generally bad to debug and refactor.

**Comment by tfausak** [comment]
We already have an entry for that: https://github.com/EdutainmentLIVE/haskell-style-guide/tree/fedd2b2ff139816a21237887535dfb2db36269bf#prefer-monads-for-building-records

**Comment by aschmois** [comment]
Oh cool, maybe we should mention that `NamedFieldPuns` can make that process easier

**Comment by tfausak** [comment]
Agreed! A "Prefer `NamedFieldPuns`" entry would be nice :)

**Comment by codygman** [comment]
One thing to note here is that positionality doesn't really matter for us since we use new types everywhere.

So in our codebase applicative style wouldn't have that downside.

**Comment by aschmois** [comment]
I think with our codebase the two main annoyances to applicative is:
- Minor refactoring (since we have newtypes refactoring is safer)
- Debugging

**Comment by codygman** [comment]
NOTE: I pretty much consider the decision to not use applicative style made and accept it, but I have no problem debating the issue since I value figuring out the right answer or best trade-offs highly. Knowledge is our work after all.

The main place we would use applicative style would be in record building and would only be unsafe if new types aren't used.

I would argue there's not much difference in debugging records built with applicative Style versus monadic Style if all fields are new types.

I do agree there is a difference in debugging applicative Style versus monadic style in the other contexts, though I'm not sure I agree it's strictly worse.

**Comment by tfausak** [comment]
I have three main problems with applicative code:

- It's easier to mess things up than comparable monadic code.
- It's harder to modify, for example by adding more checks along the way.
- As an extension of the previous point, it's harder to debug.

It's hard to come up with a compelling short examples that don't feel contrived, but imagine something like this:

``` hs
newtype Name
  = Name String
  deriving (Eq, Show, FromJSON)

newtype Age
  = Age Int
  deriving (Eq, Show, FromJSON)

data Person = Person
  { name :: Name
  , age :: Age
  } deriving (Eq, Show)

instance FromJSON Person where
  parseJSON = withObject "Person" $ \ o -> Person
    <$> o .: "name"
    <*> o .: "age"
```

Even though we use `newtype`s everywhere, it's easy to accidentally mess up the order of the field. For example, this would compile fine:

``` hs
parseJSON = withObject "Person" $ \ o -> Person
  -- oops, flipped the fields
  <$> o .: "age"
  <*> o .: "name"
```

Now you might argue that you can make the same mistake with monadic code. That's true, but it's much more obvious, assuming you've picked reasonable names for your values.

``` hs
parseJSON = withObject "Person" $ \ o -> do
  -- make sure identifier on left matches key on right
  name <- o .: "name"
  age <- o .: "age"
  pure Person { name, age }
```

Now let's say you want to change something about deserializing. Like maybe some legacy client sends the name backwards, but also sends the age negated. (Like I said, it's hard to come up with compelling, non-contrived examples.) Since the decision about what to do with the name depends on the age, you can't express that with an applicative. So if you wanted to make that change, you'd have to rewrite it in the monadic style, then make the change. Not the end of the world, but why not write it in the monadic style to begin with?

``` hs
parseJSON = withObject "Person" $ \ o -> do
  rawAge <- o .: "age"
  rawName <- o .: "name"
  pure $ if rawAge < 0
    then Person { name = reverse rawName, age = negate rawAge }
    else Person { name = rawName, age = rawAge }
```

Since debugging naturally wants to use earlier values, you run into the same type of problem there. What if you're trying to debug this so you print something out after every step? Your only choice in the applicative style is to print a constant value, but with the monadic style you can print the values too.

``` hs
parseJSON = withObject "Person" $ \ o -> do
  traceM $ "parsing object " <> show o <> " as person"
  name <- o .: "name"
  traceM $ "got name " <> show name
  age <- o .: "age"
  traceM $ "got age " <> show age
  pure Person { name, age }
```
